modules/promotion/src/Element/CouponRedemptionForm.php

```diff
@@ -92,21 +101,100 @@ public static function processForm(array $element, FormStateInterface $form_stat
       '#type' => 'textfield',
       '#title' => $element['#title'],
       '#description' => $element['#description'],
+      '#access' => !$has_coupons,
     ];
     $element['apply'] = [
       '#type' => 'submit',
       '#value' => $element['#submit_title'],
       '#name' => 'apply_coupon',
       '#limit_validation_errors' => [
+        $element['#parents'],
         array_merge($element['#parents'], ['code']),
```

Comment:
We need to pass `$element['#parents']` to ensure submit callback has proper values. Maybe we can kill second line.